forkable/forkable.go

```diff
@@ -35,9 +38,52 @@ type Forkable struct {
 
 	includeInitialLIB bool
 
+	irrChecker *irreversibilityChecker
+
 	lastLongestChain []*Block
 }
 
+type irreversibilityChecker struct {
+	answer               chan bstream.BasicBlockRef
+	blockIDServer        pbblockmeta.BlockIDServer
+	delayBetweenChecks   time.Duration
+	maxNormalLIBDistance uint64
+	lastCheckTime        time.Time
+}
+
+func (ic *irreversibilityChecker) CheckAsync(blk bstream.BlockRef, libNum uint64) {
+	if blk.Num() < libNum+ic.maxNormalLIBDistance { // only kick in when lib gets too far
+		return
+	}
+	if time.Since(ic.lastCheckTime) < ic.delayBetweenChecks {
+		return
+	}
+	ic.lastCheckTime = time.Now()
+	go func() {
+		ctx, _ := context.WithTimeout(context.Background(), ic.delayBetweenChecks)
+
+		resp, err := ic.blockIDServer.NumToID(ctx, &pbblockmeta.NumToIDRequest{
+			BlockNum: blk.Num(),
+		})
+		if err != nil {
+			zlog.Warn("forkable cannot fetch BlockIDServer (blockmeta) to resolve block ID", zap.Error(err), zap.Uint64("block_num", blk.Num()))
+		}
+		if resp.Irreversible && resp.Id == blk.ID() {
+			zlog.Debug("found irreversible block", zap.Uint64("block_num", blk.Num()), zap.String("block_id", blk.ID()))
+			ic.answer <- bstream.NewBlockRef(blk.ID(), blk.Num())
+		}
+	}()
+}
+
+func (ic *irreversibilityChecker) Found() *bstream.BasicBlockRef {
```

Comment:
Should not be a pointer, for performance-wise reason, much more efficient to copy by value.
```suggestion
func (ic *irreversibilityChecker) Found() (out bstream.BasicBlockRef, found bool) {
```